lib/spack/spack/package.py

```diff
@@ -1076,6 +1087,9 @@ def do_patch(self):
         if not self.spec.concrete:
             raise ValueError("Can only patch concrete packages.")
 
+        if not self.has_code:
+            raise InvalidPackageOpError("Can only patch a package with a URL.")
```

Comment:
Same logic applies here.  I *think* we can allow patch functions, but not patch() directives.  I could imagine someone having a use for the `patch()` instance method on a no-code package, but not the `patch()` directive (for similar reasons to `do_stage()`).  Basically, my rationale is if it doesn't hurt anything why disallow it?

I think you can just remove the extra condition here because no-source packages won't have any patches (as you disallowed them in the directive).

@becker33: thoughts?

Reply:
Interesting.

Reply:
Removed.